server/src/main/java/org/elasticsearch/search/internal/ContextIndexSearcher.java

```diff
@@ -320,4 +346,298 @@ public DirectoryReader getDirectoryReader() {
         assert reader instanceof DirectoryReader : "expected an instance of DirectoryReader, got " + reader.getClass();
         return (DirectoryReader) reader;
     }
+
+    /**
+     * iFace which implements the query timeout / cancellation logic
+     */
+    public interface Cancellable {
+
+        boolean isEnabled();
+        void checkCancelled();
+        default void checkDirReaderCancelled() {
+            checkCancelled();
+        }
+        void unsetCheckTimeout();
+    }
+
+    public static class CancellableImpl implements Cancellable {
```

Comment:
The implementation should live on the QueryPhase imo. This way you don't need to expose the `checkDirReaderCancelled` and `checkDirReaderCancelled`  in the searcher.